src/Compilers/CSharp/Portable/Symbols/SymbolWithAnnotations.cs

```diff
@@ -915,21 +906,137 @@ public override void ReportDiagnosticsIfObsolete(Binder binder, SyntaxNode synta
 
             protected override bool TypeSymbolEquals(TypeSymbolWithAnnotations other, TypeCompareKind comparison)
             {
-                var otherLazy = other as LazyNullableType;
-
+                var otherLazy = other as LazyNullableTypeParameter;
                 if ((object)otherLazy != null)
                 {
                     return _underlying.TypeSymbolEquals(otherLazy._underlying, comparison);
                 }
+                return base.TypeSymbolEquals(other, comparison);
+            }
+        }
+
+        /// <summary>
+        /// Non-nullable or null-oblivious type parameter. IsNullable is calculated
+        /// lazily to avoid cycles when binding declarations.
+        /// </summary>
+        private sealed class LazyNonNullableTypeParameter : TypeSymbolWithAnnotations
+        {
+            private readonly TypeParameterSymbol _typeParameter;
+            private readonly bool? _isNullableIfReferenceType;
```

Comment:
It feels like we should only need a `bool` here, to indicate the `NonNullTypes` interpretation of an unannotated type (`NonNullTypes = true` means oblivious, and `false` means "non-nullable").